### PR TITLE
Improve/simplify logic around weight_and_status

### PR DESCRIPTION
**Pull request by sileht**
make status_check state and desc mandatory
simply the weight_and_status
Remove duplicated cached information

### CODE REVIEW

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 7f73c2b9dae35f49a1c129dec36d254737161267

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 46088661003a9fc1ee065ff8cde385eabbc4abd7

**Comment by sileht** [comment]
Pull-request updated, HEAD is now adab7f23d7f4656b743c8ee812f21291ded1c51f

**Comment by sileht** [comment]
Pull-request updated, HEAD is now e39a893ec7eae5f58ccfc70a2e7e32555fa82d7e

File: mergify_engine/gh_pr_fullifier.py [hunk 1]
```diff
@@ -103,78 +103,90 @@ def disabled_by_rules(pull, **extra):
...
+# unstable: branch up2date (if strict: True) and not required status
+#           checks are failure or pending
+# clean: branch up2date (if strict: True) and all status check OK
+#
+# https://platform.github.community/t/documentation-about-mergeable-state/4259
+# https://github.com/octokit/octokit.net/issues/1763
+
+
+# Use enum.Enum and make is serializable ?
+class MergifyState(object):
```
Comment (jd) [inline mergify_engine/gh_pr_fullifier.py]: use `enum.Enum`

File: mergify_engine/gh_pr_fullifier.py [hunk 2]
```diff
@@ -103,78 +103,90 @@ def disabled_by_rules(pull, **extra):
...
-        # Maybe clean soon, or maybe this is the previous run
-        # selected PR that we just rebase
-        weight = 10
-        status_desc = "Waiting for CI success"
+        # Maybe clean soon, or maybe this is the previous run selected PR that
+        # we just rebase, or maybe not. But we set the mergify_state to
+        # ALMOST_READY to ensure we do not rebase multiple pull request in //
+        mergify_state = MergifyState.ALMOST_READY
+        github_state = "pending"
+        github_desc = "Waiting for CI success"
```
Comment (jd) [inline mergify_engine/gh_pr_fullifier.py]: Could we change that to `"Waiting for status checks success"` which is more inline about what we wait (CI is more like a corner case of status check).

File: mergify_engine/gh_pr_fullifier.py [hunk 3]
```diff
@@ -103,78 +103,90 @@ def disabled_by_rules(pull, **extra):
...
                            "branch changes soon")
-        elif pull.mergify_engine["combined_status"] == "pending":
-            weight = 5
-            status_desc = "Waiting for CI success"
         else:
-            weight = -1
-            status_desc = "Waiting for CI success"
+            mergify_state = MergifyState.NOT_READY
+            github_state = "pending"
+            github_desc = "Waiting for CI success"
```
Comment (jd) [inline mergify_engine/gh_pr_fullifier.py]: ditto

use a variable  or class or something to DRY :)

File: mergify_engine/gh_pr_fullifier.py [hunk 4]
```diff
@@ -103,78 +103,90 @@ def disabled_by_rules(pull, **extra):
...
-    return (weight, status_desc)
-
-
-def compute_weight(pull, **extra):
-    # TODO(sileht): Don't store this twice
-    return pull.mergify_engine["weight_and_status"][0]
-
+        mergify_state = MergifyState.NOT_READY
+        github_state = "pending"
+        github_desc = "Waiting for CI success"
```
Comment (jd) [inline mergify_engine/gh_pr_fullifier.py]: ditto

File: mergify_engine/gh_pr_fullifier.py (hunk 1 above)
Comment (sileht) [reply mergify_engine/gh_pr_fullifier.py]: That's planned, but I need first to improve how we serialize/deserialize thing is redis

File: mergify_engine/gh_pr.py [hunk 5]
```diff
@@ -156,11 +150,14 @@ def monkeypatch_github():
     p.mergify_engine_github_post_check_status = \
         mergify_engine_github_post_check_status
 
+    p.milestone_is_set = property(
```
[the remark this answers is not in the document]
Comment (sileht) [reply mergify_engine/gh_pr.py]: done

File: mergify_engine/gh_pr_fullifier.py (hunk 2 above)
Comment (sileht) [reply mergify_engine/gh_pr_fullifier.py]: done

File: mergify_engine/gh_pr_fullifier.py (hunk 3 above)
Comment (sileht) [reply mergify_engine/gh_pr_fullifier.py]: done

File: mergify_engine/gh_pr_fullifier.py (hunk 4 above)
Comment (sileht) [reply mergify_engine/gh_pr_fullifier.py]: done

**Comment by sileht** [comment]
Pull-request updated, HEAD is now f9a2e2373e77cc285100bd02ea6f1553050748ad

**Comment by sileht** [comment]
Pull-request updated, HEAD is now b6e93d58bb054893b8ab0dd88d090849d248e160

**Pull request merged**